### ISSUE TITLE
fmap.hdr.session or fmap.session?

### ISSUE THREAD

**Issue by jooh** [automaticanalysis/automaticanalysis#214]
I'm trying to resolve a merge conflict in my fork. It looks like there may have been a change in expected syntax. Here is what is in the current AA master:

https://github.com/automaticanalysis/automaticanalysis/blob/53ef81ec592b24cb1981261c0ec57b813a1a593c/aa_engine/aas_processBIDS.m#L286-L293

And here is my fork:

https://github.com/jooh/automaticanalysis/blob/efbff91382d605a07dd13ccbe5746a1b4702e97d/aa_engine/aas_processBIDS.m#L246-L253

It looks like the 'session' field now nests under the 'hdr' field. Is this the correct behaviour? Does anyone have an example BIDS script where IntendedFor is used, so we can test that this works as intended? Going to assign @jones-michael-s since that's where the change came from.

See also #187

**Comment by tiborauer** [comment]
The 7T BIDS dataset (in /imaging/local/software/AA/test_resources/7t_trt) is such a test dataset.
You can try the old BIDS7T example (I found the latest in https://github.com/automaticanalysis/automaticanalysis/tree/22c4827ed1be5f46b3c26c75f9a7969e0a721e50).

**Comment by jones-michael-s** [comment]
Hi Johan,

It's true I recently modified aas_processBIDS.m, but not this area of the code.

Does anyone else recognize the code JC lists above?

**Comment by tiborauer** [comment]
Considering https://github.com/automaticanalysis/automaticanalysis/blob/dc9f500dab542008cc361eae0b578e548bfa7685/aa_modules/aamod_fieldmapfromnifti.m#L20-L27, it is not the correct behaviour any more. Although it has been before. I am to blame for the inconsitency, because I have not changed the aas_processBIDS in that commit.

**Comment by jooh** [comment]
> It's true I recently modified aas_processBIDS.m, but not this area of the code.

@jones-michael-s git blame shows that you were the last to edit these lines. Am I missing something? Here is the commit:

https://github.com/automaticanalysis/automaticanalysis/commit/f202cef65b5a2b9a57d30bb32bcda1da50882f15#diff-1a5aef641251e88cd90e1287d4814744L247

> it is not the correct behaviour any more. Although it has been before.

@tiborauer Those lines in aamod_fieldmapfromnifti appear to look for 'session', not 'hdr.session'. So are we going with 'session' and changing processBIDS?

> The 7T BIDS dataset (in /imaging/local/software/AA/test_resources/7t_trt) is such a test dataset.

Sounds good! It shouldn't be too difficult to re-introduce that example. So this used to run on empty raw data files from [bids-examples](https://github.com/bids-standard/bids-examples)? Or from a version of that dataset with real files? Do you know if the dataset is available in a permanent location somewhere so we can use aa_downloaddemo to grab them on demand (without introducing a git-annex / datalad dependency)?

**Comment by tiborauer** [comment]
I suspect @jones-michael-s was working on an outdated version and did not performed a rebase before PR.

Anyway, I think we should go with 'session' rather than 'hdr.session', because it is supported by aamod_fieldmapfromnifti. 

The script requires real dataset. The *fromnifti stages may run on the empty one, too.
The 7T TRT dataset is available on [OpenNEURO](https://openneuro.org/datasets/ds001168/versions/00003/download). However, one needs AWS CLI to download it, and I am not sure whether you can download selectively.

**Comment by jones-michael-s** [comment]
I made sure my fork was up-to-date with the main repo by doing a pull (to my local machine) and a push (to my fork) before opening the PR. Is that equivalent to "performed a rebase"?

**Comment by tiborauer** [comment]
push-pull is between your fork and your local machine
If you do not re-merge (i.e. reverse PR) main repo to your fork, then you might miss out changes made by others' PR.

**Comment by jones-michael-s** [comment]
I should have been more clear: I pulled from my fork's **upsteam** (which I previously defined as the main aa repo). Then I pushed to my fork...

**Comment by jones-michael-s** [comment]
Update: Okay, I think I see what happened. My fork is indeed up-to-date with the main repo -- I did the pull/push then to make sure of this before making my changes to aas_processBIDS.m in my previous PR and that all worked properly.

However, instead of _editing_ aas_processBIDS.m, I _replaced_ it. Unfortunately, this new version was based on the pre-philips merge. Ergo, my PR clobbered the phasediff code listed above. Apparently, none of our testing ever exercised this block.

Since there is currently a new PR open, what is the easiest way to fix this? Can you just replace (or rewind or whatever the git term is) aas_processBIDS.m with the version before my PR into the current PR? After the current PR is merged, I can submit a new PR with my aas_processBIDS.m changes again (I promise to edit the proper version of the code this time...).

Mea culpa,
Mike

**Comment by tiborauer** [comment]
That is a very plausible explanation. I think a revert is, indeed, the  easiset solution unless @jooh can suggest something more clever. :)

**Comment by tiborauer** [comment]
There are two commits/merges in my repo with similar entries ([this](https://github.com/tiborauer/automaticanalysis/commit/53ef81ec592b24cb1981261c0ec57b813a1a593c) and [this](https://github.com/tiborauer/automaticanalysis/commit/dc29787db37dc14500fc2663f11adc5137e2351b)). Which one I should revert? Does it matter? Or shall I simply the file from the commit just before the changes and add as a restore?

**Comment by jones-michael-s** [comment]
I submitted the PR then submitted an update to the PR while it was open which would account for the two commits. I suggest taking the version from before both of those changes.

Thanks,
M

**Comment by tiborauer** [comment]
Done it.

**Comment by jooh** [comment]
Good to hear things make sense again! Thanks for looking into this.

Regarding merge / rebase, that's not really the issue here. Might be worth reading up on the docs for how these two differ. They're basically different ways of encoding the history of the commits when you combine commits from two branches that may have been developed in parallel. With merge all branch commits are kept in history (ie, the history is like a tree, which is correct but can be tricky to work with), while with rebase you rewrite history so that the changes in the branch are put on top of the current state of the master branch (so history becomes like a line, which isn't how it actually happened at the time, but makes for a cleaner log). Because rebase re-writes history it can break repos for other people, so best done on your own local branches before submitting a pull request. Personally I never use it but it seems popular?

Regarding how to fix, I think @jones-michael-s might be able to [check out Tibor's pull request](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/checking-out-pull-requests-locally), make his changes again, and push it back to the same PR. Or perhaps easier, submit them as a PR to Tibor's fork of the repo, and they will get added to Tibor's PR when he merges them.

It would be a shame to lose those changes and then have to run the test suite all over again when they're added back in. I am also planning to push in a few small changes from our end before running the tests.

**Comment by jones-michael-s** [comment]
I'll try to fix this in current PR (fingers crossed). 

I'm using [https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/committing-changes-to-a-pull-request-branch-created-from-a-fork](url) as a guide. This seems to suggest I need permission from @tiborauer to push to his fork.

I'll be testing my code locally today; hopefully ready to push this afternoon if all goes well.

**Comment by tiborauer** [comment]
I also suggest to send it as a PR to my fork.

**Comment by jones-michael-s** [comment]
Okay. I'll try that.

**Comment by jones-michael-s** [comment]
Okay, just opened a PR on @tiborauer fork.

Hopefully this is (finally!) correct.

**Comment by jones-michael-s** [comment]
So I was looking into how I messed up the previous PR (the one that clobbered the phasediff changes Johan discovered) and I found something I don't understand.

Before I opened the PR, I did a pull into my repo to make sure any changes from the main repo were merged -- just the standard recommended good git practice.

This indeed updated my master branch correctly. Importantly, it updated aas_processBIDS.m -- my repo had an out-of-date version.

I then did a merge from my master into the branch I had created to make my changes (which I called "BIDS_extensions"). That is, I did:

$ git checkout BIDS_extensions
$ git merge master

It turns out this merge didn't do anything. It simply returned "Already up to date". 

The odd thing is if I now do a git diff master, git lists a bunch of differences between the version of aas_processBIDS.m in BIDS_extensions and the version in master (as it should). And yet, repeated attempts to merge do nothing.

I don't understand this. I would expect if git diff finds differences in a file between two branches, a merge would necessarily combine them. At worst, it might generate a merge conflict, but it shouldn't ignore them.

Can anyone explain what I'm doing wrong?

M

**Comment by tiborauer** [comment]
I am not sure I undestand your workflow. Mine is this:
1. On GitHub, I merge (PR) central repo master branch into my feature branch. I usually flags up some conflicts. 

2. I use the command line tools and an editor (VS Code) to resolve conflicts. When doing so, I make sure I do not cause uninteded changes by accepting changes based on out-of-date version.

3. I push resolves into my feature branch, which automatically update stratus of the PR to "merged".

4. I make a PR to the central repo master branch.

**Comment by tiborauer** [comment]
Also, my feature branch is based on central repo master branch or my master branch aligned with the central repo master branch:

1.a: git checkout -b <feature> <remote>/<base branch>
1.b: git checkout -b <feature> origin/<base branch>
2. editing and committing
3. git push -u origin <feature>

**Comment by jones-michael-s** [comment]
I figured out why my PR went sideways. 

I'll close this issue.
